torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -812,7 +979,28 @@ void SegmentCandidateFinder::findSegments() {
     }
   }
 
+  // Set heuristics in case single reduction kernels were left out
+  for (auto g : groups()) {
+    g->setHeuristic(deriveHeuristic(g));
+  }
+
+  // Scalar merge iteration, merge all scalar edges
+  bruteForceMerge([this](SegmentedEdge* edge) {
+    return edge->val->isScalar() && codeGenSupportedMerge(edge);
+  });
```

Comment:
@csarofeen Wanted to see your suggestions before merging this PR. To fix some tests like layer_norm_backward, I had to run a brute force iteration before the actual merging to make sure that all scalar edges are merged into some groups. It has to be the brute force version of merging to guaranteed that no scalar edges are left behind in the end.

Also added assertion before actual merging and user can fix this assertion failure by avoiding multiple uses of a computed scalar. 

I guess in longer term if we officially not support multiple uses of scalar expressions, the initial scalar merging iterations can be much more lightweight but I'm not sure if that's reasonable.

Reply:
Why don't we just include all scalar computations in any graph that uses them? Scalar computations should not impact the result of `codeGenSupportedMerge`. So simply expanding to include any expression resulting in a scalar that the group uses should be fine. It may even replicate the computation of the scalar which also should be fine.